onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#if (defined(_M_AMD64) && !defined(_M_ARM64EC)) || defined(_M_IX86) || defined(__x86_64__) || defined(__i386__) || !defined(DISABLE_CONTRIB_OPS)
+
+#include "core/optimizer/avx2_weight_s8_to_u8.h"
+
+#include <algorithm>
+
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+
+/**
+ * @brief Convert the source int8_t TensorProto to a uint8_t one if the tensor contains
+ *        value outside of [-64, 63]
+ * @param src           The source tensor, must be type int8_t
+ * @param dst           An empty tensor, will contain the converted tensor data
+ * @param extern_path   In case the source tensor data is external, need to provide
+ *                      data path
+ * @param force         Perform conversion even when tensor values within [-64, 63]
+ * @return              Whether the conversion happened.
+*/
+static inline
+bool Int8TensorProto2Uint8(const ONNX_NAMESPACE::TensorProto& src, ONNX_NAMESPACE::TensorProto& dst,
+    const Path& extern_path, bool force = false) {
+
+    dst.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_UINT8);
+    dst.set_name(src.name() + "_s8_2_u8");
+    dst.mutable_dims()->CopyFrom(src.dims());
+
+    // TODO(fuchen): too many copies!
+    //
+    // Here we do two memory copies: Proto -> Initializer -> Proto.
+    // Ideally we only do 1 copy, just iterate the source data, and write directly to the
+    // dst raw buffer.
+    // Unfortunately iterating the source data is complicated, the data maybe in external
+    // file, a raw buffer, or a repeated field depending on the data type.  UnpackTensor()
+    // already contains some of these logic and is closest to what we need. But it does
+    // not handle external data. Write our own code here means copy the logic of
+    // TensorProtoToTensor(), a violation of DRY principle.
+
+    Initializer temp(src, extern_path);
+    int8_t* p = temp.data<int8_t>();
+    bool should_convert = false;
+    for (int i = 0; i < temp.size(); i++) {
+      if (*p < -64 || *p > 63) {
+        should_convert = true;
+      }
+      *p ^= 0x80;
+      p++;
+    }
+    if (force || should_convert) {
+      dst.set_raw_data(temp.data<int8_t>(), temp.size());
+      return true;
+    }
+    return false;
+}
+
+/**
+ * @brief If the op_node has an uint8_t const weight tensor, convert it to int8_t
+ * @param graph 
+ * @param op_node 
+ * @param weights_idx
+ * @param weight_zp_idx
+ * @return true when conversion happened.
+*/
+static bool ConvertS8WeightToU8(Graph& graph, Node& op_node, size_t weights_idx, size_t weight_zp_idx) {
+  auto& input_defs = op_node.MutableInputDefs();
+  if (input_defs.size() < std::max(weights_idx, weight_zp_idx) + 1) {
+    // TODO(fuchen): under what condition weight zp could be null? or not constant?
```

Comment:
>zp could be null [](http://example.com/codeflow?start=49&length=16)

zp can be null if using default value(0).

Reply:
fixed